src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  actions.forEach((action) => {
+    const previousState =
+      states.length > 0 ? states[states.length - 1] : { ...state };
+    let newState = { ...previousState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
```

Comment:
[CODE STYLE]: It is good practice to include a `default` case in a `switch` statement for error handling. In this case, you have included the default case, which is good, but consider throwing an error or logging a message if an unknown action type is encountered to aid in debugging.